rfd/0067-database-access-aws-redis

```diff
@@ -0,0 +1,260 @@
+For instance, a database user with passwords `v7` and `v8` is getting a
+password rotation to use `v8` and `v9`, after `v9` is created in the secret
+store. `v8` is a valid password while the password change is propagating or
+after. (From another point of view, `v8` is the current password and `v9` is
+the next.)
+
+When a client tries to connect to ElastiCache and MemoryDB Redis servers, the
+database service automatically logins Teleport managed database users by
+sending `auth` commands with the stored passwords to the Redis server. For
+database users not managed by Teleport, users can manually input `auth`.
+
+### Secret store
+
+#### Secret store - Interface
+
+For ElastiCache/MemoryDB Redis usage, it is natural to use an AWS native
+solution (aka [AWS Secrets Manager](https://aws.amazon.com/secrets-manager/)).
+Though other secret stores could be supported in the future. <sup>[2]<sup>
+
+Here is a sample interface for managing secrets using the secret store.
+```
+// SecretValue is the secret value.
+type SecretValue struct {
+    // Key is the key path of the secret.
+    Key string
+    // Value is the value of the secret.
+    Value string
+    // Version is the version of the secret value.
+    Version string
+    // Created is the creation time of this version.
+    Created time.Time
+}
+
+// PutValueRequest defines the request to create a new secret version.
+type PutValueRequest struct {
+    // Key is the key path of the secret.
+    Key string
+    // Value is the value of the secret to set. A random value is generated if
+    // empty.
+    Value string
+    // RandomValueLength is the length of the random value to generate.
+    RandomValueLength int
+    // CurrentVersion is the version of the current secret value. If set, it
+    // will be used for test-and-set to ensure only one caller succeeds for the
+    // same version.
+    CurrentVersion string
+}
+
+// Secrets defines an interface for managing secrets. A secret consists of a
+// key path and a list of versions that hold copies of current or past secret
+// values.
+type Secrets interface {
+    // PutValue creates a new secret version for the secret. The secret is also
+    // created if it does not exist.
+    PutValue(req PutValueRequest) (*SecretValue, error)
+
+    // GetValue returns the secret value for provided version. Besides version
+    // string returned from PutValue, two specials versions "CURRENT" and
+    // "PREVIOUS" can also be used to retrieve the current and previous
+    // versions respectively. If the version is empty, "CURRENT" is used.
+    GetValue(key string, version string) (*SecretValue, error)
+
+    // Delete deletes the secret for the provided path.
+    Delete(key string) error
+}
+```
+It is up to the caller to call `PutValue` periodically to rotate the secret.
+
+#### Secret store - AWS Secrets Manager
+
+AWS Secrets Manager is used to store passwords for ElastiCache and MemoryDB
```

Comment:
One thing that could be missing here is a some price estimation. AWS secret manager is not free and depending on how many secrets we will create/API calls we will make a price may be "visible" for some clients on theirs AWS bill. As this will be a mandatory feature for AWS Redis the cost of AWS secret manager will be at that point an additional cost of the Teleport integration.